File: DMS/apps/mobile-rn/src/request_signer.ts
import { createHmac } from 'crypto';

export interface CanonicalRequestParts {
  method: string;
  path: string;
  query: string;
  body: string;
  timestamp: string;
  nonce: string;
}

/**
 * Build a canonical request string byte-identical to the server TypeScript code.
 */
export function canonicalRequestString(parts: CanonicalRequestParts): string {
  const method = parts.method.toUpperCase();
  const path = normalizePath(parts.path);
  const query = normalizeQuery(parts.query);
  const bodyHash = createHmac('sha256', Buffer.alloc(32))
    .update(parts.body)
    .digest('hex');

  return [
    method,
    path,
    query,
    bodyHash,
    parts.timestamp,
    parts.nonce,
  ].join('\n');
}

function normalizePath(path: string): string {
  let normalized = path.replace(/\/+/g, '/');
  if (!normalized.startsWith('/')) {
    normalized = '/' + normalized;
  }
  if (normalized.length > 1 && normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}

function normalizeQuery(query: string): string {
  const cleaned = query.startsWith('?') ? query.slice(1) : query;
  if (!cleaned) {
    return '';
  }

  const params = cleaned.split('&').filter(Boolean);
  const pairs: Array<[string, string]> = params.map((p) => {
    const eqIndex = p.indexOf('=');
    if (eqIndex === -1) {
      return [decodeURIComponent(p), ''] as [string, string];
    }
    return [
      decodeURIComponent(p.slice(0, eqIndex)),
      decodeURIComponent(p.slice(eqIndex + 1)),
    ] as [string, string];
  });

  pairs.sort((a, b) => {
    const keyCompare = a[0].localeCompare(b[0]);
    if (keyCompare !== 0) return keyCompare;
    return a[1].localeCompare(b[1]);
  });

  return pairs
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
    .join('&');
}

/**
 * Sign canonical headers with a client-side key.
 */
export function signRequest(parts: CanonicalRequestParts, secretKeyHex: string): string {
  const canonical = canonicalRequestString(parts);
  const key = Buffer.from(secretKeyHex, 'hex');
  return createHmac('sha256', key).update(canonical).digest('hex');
}

File: DMS/apps/mobile-flutter/lib/crypto/request_signer.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

class CanonicalRequestParts {
  final String method;
  final String path;
  final String query;
  final String body;
  final String timestamp;
  final String nonce;

  CanonicalRequestParts({
    required this.method,
    required this.path,
    required this.query,
    required this.body,
    required this.timestamp,
    required this.nonce,
  });
}

class RequestSigner {
  /**
   * Build a canonical request string byte-identical to the server TypeScript code.
   */
  static String canonicalRequestString(CanonicalRequestParts parts) {
    final method = parts.method.toUpperCase();
    final path = _normalizePath(parts.path);
    final query = _normalizeQuery(parts.query);
    
    // Body hash is hmac-sha256 with 32 zero bytes as key
    final zeroKey = List<int>.filled(32, 0);
    final hmac = Hmac(sha256, zeroKey);
    final bodyHash = hmac.convert(utf8.encode(parts.body)).toString();

    return [
      method,
      path,
      query,
      bodyHash,
      parts.timestamp,
      parts.nonce,
    ].join('\n');
  }

  /**
   * Sign canonical headers with a client-side key.
   */
  static String signRequest(CanonicalRequestParts parts, String secretKeyHex) {
    final canonical = canonicalRequestString(parts);
    final key = _hexDecode(secretKeyHex);
    final hmac = Hmac(sha256, key);
    return hmac.convert(utf8.encode(canonical)).toString();
  }

  static String _normalizePath(String path) {
    var normalized = path.replaceAll(RegExp(r'/+'), '/');
    if (!normalized.startsWith('/')) {
      normalized = '/' + normalized;
    }
    if (normalized.length > 1 && normalized.endsWith('/')) {
      normalized = normalized.substring(0, normalized.length - 1);
    }
    return normalized;
  }

  static String _normalizeQuery(String query) {
    final cleaned = query.startsWith('?') ? query.substring(1) : query;
    if (cleaned.isEmpty) {
      return '';
    }

    final params = cleaned.split('&').where((p) => p.isNotEmpty).toList();
    final pairs = params.map((p) {
      final eqIndex = p.indexOf('=');
      if (eqIndex == -1) {
        return MapEntry(Uri.decodeComponent(p), '');
      }
      return MapEntry(
        Uri.decodeComponent(p.substring(0, eqIndex)),
        Uri.decodeComponent(p.substring(eqIndex + 1)),
      );
    }).toList();

    // Sort parameters alphabetically
    pairs.sort((a, b) {
      final keyCompare = a.key.compareTo(b.key);
      if (keyCompare != 0) return keyCompare;
      return a.value.compareTo(b.value);
    });

    return pairs
        .map((p) => '${Uri.encodeComponent(p.key)}=${Uri.encodeComponent(p.value)}')
        .join('&');
  }

  static List<int> _hexDecode(String hex) {
    final result = <int>[];
    for (var i = 0; i < hex.length; i += 2) {
      result.add(int.parse(hex.substring(i, i + 2), radix: 16));
    }
    return result;
  }
}
